### ISSUE TITLE
DeepSpeed resume fails for multiple nodes

### ISSUE THREAD

**Issue by sonodasong** [mosaicml/composer#2326]
** Environment **

An environment with 2 nodes, each node has 1 GPU, shared file system.
deepspeed==0.8.3
mosaicml==0.14.1

** To reproduce

1. Launch the job in ```node0```
```CUDA_VISIBLE_DEVICES=0 composer --world_size 2 --node_rank 0 --master_addr node0```
and in ```node1```
```CUDA_VISIBLE_DEVICES=0 composer --world_size 2 --node_rank 1 --master_addr node0```
2. Once checkpoints saved, stop the job.
There should be ```latest-rank0.pt.tar``` and ```latest-rank1.pt.tar``` in the ```save_folder```
3. Specify ```load_path``` to ```save_folder/f'latest-rank{composer.utils.dist.get_global_rank()}.pt.tar'``` in ```Trainer```
4. Launch the resumed job  in ```node0```
```CUDA_VISIBLE_DEVICES=0 composer --world_size 2 --node_rank 0 --master_addr node0```
and in ```node1```
```CUDA_VISIBLE_DEVICES=0 composer --world_size 2 --node_rank 1 --master_addr node0```

Then there will be error in ```node1```:
```FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmp3o00dg6a/checkpoint/composer_states.pt'```

## Expected behavior

The resume should work without errors.

## Proposed fix

The reason is because only (global) rank 0 saves ```composer_states.pt```.
https://github.com/mosaicml/composer/blob/7da93f819c4d5f52ce5ccfef0f39a7f836cb0d08/composer/utils/checkpoint.py#L530-L535
So only ```latest-rank0.pt.tar``` has ```composer_states.pt```.

In order for the other (global) ranks to access ```composer_states.pt```, all (global) ranks should share the same ```node_checkpoint_folder``` as in
https://github.com/mosaicml/composer/blob/7da93f819c4d5f52ce5ccfef0f39a7f836cb0d08/composer/utils/checkpoint.py#L204-L205

However, ```_get_local_rank_zero_path(tempdir)``` as in
https://github.com/mosaicml/composer/blob/7da93f819c4d5f52ce5ccfef0f39a7f836cb0d08/composer/utils/checkpoint.py#L237-L243
cannot make every (global) rank share the same ```node_checkpoint_folder```.
Instead, only local ranks on the same node share the same ```node_checkpoint_folder```.

The correct way should be use ```_get_global_rank_zero_path(tempdir)```:
```python
def _get_global_rank_zero_path(path: Optional[str]) -> str:
    """Broadcasts the ``path`` from the (GLOBAL) rank zero to all (GLOBAL) ranks."""
    paths = dist.all_gather_object(path)
    rank_zero_path = paths[0]
    assert rank_zero_path is not None, 'rank zero provides the path'
    return rank_zero_path
```

In summary, only (global) rank 0 saves ```composer_states.pt```.
However, the code assumes every local rank 0 saves ```composer_states.pt```.

**Comment by bcui19** [comment]
Hey @sonodasong is it possible to download the checkpoint on `node1` as well? 

Alternatively, for composer, we've more typically used FSDP. Is there a requirement to use deepspeed?

**Comment by sonodasong** [comment]
> Hey @sonodasong is it possible to download the checkpoint on `node1` as well?

There is no problem for download on `node1`.

> Alternatively, for composer, we've more typically used FSDP. Is there a requirement to use deepspeed?

We are exploring ways to pretrain transformers faster, and open to techniques that can help to achieve this.

Thanks.

**Comment by bcui19** [comment]
@sonodasong hm... If it's possible to download the model on `node0` and `node1` then does resumption work?

Makes sense, at mosiac we have open sourced a repository for LLM-pretraining:

https://github.com/mosaicml/llm-foundry/

Typically, for this repo, we use FSDP as the distributed backend. We're able to achieve 50-60% MFU (https://github.com/mosaicml/llm-foundry/tree/7abae0d079e8c99a8f1eb73861b65e6766e45018/scripts/train/benchmarking). Let me know if you have any questions. 

I'm going to close the issue for now, but feel free to re-open if the above doesn't satisfy your questions.

**Comment by sonodasong** [comment]
> @sonodasong hm... If it's possible to download the model on `node0` and `node1` then does resumption work?

No, because `node0` and `node1` download to different `extracted_checkpoint_folder`.
